### PR TITLE
Add table join implementation

### LINKED ISSUE #11010: Exception when stream attribute of type string being used as a method param in selector
**Description:**
$subject in streaming queries.

**Steps to reproduce:**
The following query will fail when the type of `inputStream.name` is a string.

```ballerina
    forever {
        from inputStream window lengthWindow(1)
        select getDefaultName(inputStream.name) as teacherName, getDefaultAge(inputStream.age, 2) as age
        => (TeacherOutput[] emp) {
            foreach e in emp {
                outputStream.publish(e);
            }
        }
    }
```

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
Purpose of this PR is to add table join desugar implementation and fix #11010.

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 1]
```diff
@@ -245,8 +256,21 @@ public void visit(BLangStreamingQueryStatement queryStmt) {
             isInJoin = true;
             BLangStreamingInput streamingInput = (BLangStreamingInput) joinStreamingInput.getStreamingInput();
             createStreamAliasMap(streamingInput);
-            rhsStream = (BLangSimpleVarRef) joinStreamingInput.getStreamingInput().getStreamReference();
-            lhsStream = (BLangSimpleVarRef) queryStmt.getStreamingInput().getStreamReference();
+            // If the node kind is INVOCATION, considering that as a table join.
+            if (queryStmt.getStreamingInput().getStreamReference().getKind() == NodeKind.INVOCATION) {
```
Comment (mohanvive) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: What if there is a variable global variable (which is the table) defined and it is used when joining with streams? I don't think considering it only as Invocation is wrong.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 2]
```diff
@@ -1651,8 +1845,13 @@ private BLangInvocation refactorInvocationWithIndexBasedArgs(BVarSymbol mapVarSy
...
                     BLangExpression fieldAccessExpr = createMapVariableIndexAccessExpr(mapVarSymbol, expr);
-                    expressionList.add(createCheckedExpr(generateConversionExpr(fieldAccessExpr,
-                            expr.type, symResolver)));
+                    if (expr.type.tag == TypeTags.STRING) {
+                        // <string> is not a checked cast
+                        expressionList.add(generateConversionExpr(fieldAccessExpr, expr.type, symResolver));
+                    } else {
+                        expressionList.add(createCheckedExpr(generateConversionExpr(fieldAccessExpr,
+                                expr.type, symResolver)));
+                    }
```
Comment (mohanvive) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: Then we have to fix in another one place as well right (where the output events are created)?

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/streaming/BallerinaStreamsV2TableJoinTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import org.ballerinalang.model.values.BMap;
+import org.ballerinalang.model.values.BValue;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * This contains methods to test streams and table join in Ballerina Streaming V2.
+ *
+ * @since 0.980.0
```
Comment (mohanvive) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/streaming/BallerinaStreamsV2TableJoinTest.java]: Fix the version value in the comment.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java [hunk 4]
```diff
@@ -147,16 +151,22 @@
...
+    private static final String SET_JOIN_PROPERTIES_METHOD_NAME = "setJoinProperties";
+    private static final String SET_RHS_METHOD_NAME = "setRHS";
+    private static final String SET_LHS_METHOD_NAME = "setLHS";
     private static final String EVENT_DATA_VARIABLE_NAME = "data";
     private static final String EVENT_TYPE_VARIABLE_NAME = "eventType";
     private static final String BUILD_STREAM_EVENT_METHOD_NAME = "buildStreamEvent";
     private static final String STREAM_SUBSCRIBE_METHOD_NAME = "stream.subscribe";
     private static final String VAR_FOREACH_VAL = "val";
     private static final String VAR_OUTPUT_EVENTS = "outputEvents";
+    private static final String VAR_RESULTS_TABLE = "resultsTable";
```
Comment (mohanvive) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: What will happen if there is any global variable defined in the above variable names? We should have some unique value (as similar how lambda functions are named).

File: tests/ballerina-unit-test/src/test/resources/test-src/streaming/streamingv2-table-join-test.bal [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+function addToGlobalEventsArray(StockWithPrice s) {
+    globalEventsArray[index] = s;
+    index = index + 1;
+}
+
+function clear() {
+    globalEventsArray = [];
+    index = 0;
+}
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/resources/test-src/streaming/streamingv2-table-join-test.bal]: Shall we add a new line to the end of the file?

Comment (grainier) [reply tests/ballerina-unit-test/src/test/resources/test-src/streaming/streamingv2-table-join-test.bal]: Fixed in 3d44c2e

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/streaming/BallerinaStreamsV2TableJoinTest.java (hunk 3 above)
Comment (grainier) [reply tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/streaming/BallerinaStreamsV2TableJoinTest.java]: Fixed in 85263fd

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java (hunk 4 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: Fixed in 3a4444a

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java (hunk 2 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: Fixed in dc79d05

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java (hunk 1 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java]: Fixed in b7e0182

**Pull request merged**